### ISSUE TITLE
End-to-end distributed tracing support

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#939]
### Problem
Azure Functions has integration with Application Insights and supports end-to-end tracing for various binding types. However, there is no support for end-to-end tracing in Durable Functions.

### Proposed Solution
Durable Functions should support end-to-end tracing and associated visualization in Application Insights.

* A user should be able to see the order of activity function calls from an orchestration, including any sub-orchestrations
* User logs should be able to correlate with these traces
* Ideally this works with external events as well
* Durable HTTP APIs should correctly flow end-to-end trace IDs
* Calls to external services supported by App Insights should flow the correct correlation info
* Should be compatible with "App Map" visualizations

The first step to supporting this is to flow correlation information through the DTFx libraries. Once that is done, it should be surfaced to the Azure Functions / Durable Functions layers. The DTFx work is currently being tracked here: https://github.com/Azure/durabletask/pull/261

### Workarounds
We do have tracing to Application Insights which includes the orchestration instance ID across orchestrator and activity function traces. However, it is not correlated with user logs. It's also not possible to directly see the call chain between orchestrations and specific activity function invocations. It also doesn't work with App Insights visualization features.

**Comment by idg10** [comment]
One aspect of this that you've not mentioned in the bullet list in your proposed solutions is how Eternal Orchestrations https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-eternal-orchestrations should fit into this.

My worry would be that once Durable Functions integrate with Application Insights in the same way that other functions triggers do today, all work started from an Eternal Orchestration would appear to be part of the same massive operation.

If possible, what I'd like to see is each iteration of an Eternal Orchestration effectively appearing as a distinct Operation in Application Insights. So each time you `ContinueAsNew` you start a new Operation for monitoring purposes.

**Comment by TsuyoshiUshio** [comment]
Hi @idg10 , 
Thank you for the comment. 
Durable Functions Distributed Tracing support won't be the same as the other functions.  It will be a separate tracing mechanism.  Since, some people need correlation in Eternal Orchestration scenario, however, now I know some people want start from new when it is ContinueAsNew. Which feature does it fit your use case? 

* Having a configuration to make ContineAsNew starts new tracing for each time.
* Having a configuration to make a specific Orchestrator ContinueAsNew start new tracing for each tiem. But other's is not. 

Or would you prefer other solution? 
@cgillum  If you have any comment to that spec, please leave a comment.

**Comment by GavinOsborn** [comment]
@cgillum thank you for this, your proposed solution sounds exactly like what I am hoping for.
If there is a build of any of this that I could play with I would be happy to help provide feedback.

**Comment by GavinOsborn** [comment]
Is this to be a v3 feature or will this be back-ported into a v2 release also?

**Comment by cgillum** [comment]
@GavinOsborn sorry for the late response. Most likely we will target Functions 3.0 for this because .NET Core 3 contains a lot of improvements in this area that we intend to take advantage of.

**Comment by MedAnd** [comment]
@cgillum - any update on this feature? Would be great to have an end to end sample & documentation for how to correlate across HttpTrigger, Durable Functions + Activities, Service Bus messages etc?

Also if using .Net Core 3+, would be good to clarify if one should use Application Insights libraries directly or the new OpenTelemetry support?

**Comment by cgillum** [comment]
@MedAnd yes, we're making progress on this work, I believe [this PR](https://github.com/Azure/azure-functions-durable-extension/pull/1270) contains the latest. @TsuyoshiUshio should be able to answer your questions about Application Insights vs. OpenTelemetry.

**Comment by TsuyoshiUshio** [comment]
Hi @MedAnd 

We are currently support AppInsights. Both W3C trace context and HttpCorrelation Protocol. 
https://devblogs.microsoft.com/aspnet/observability-asp-net-core-apps/  We cuspport .NetCore 3+.

For the OpenTelemetry, our telemetry model is created as open-closed principal, so it is easy to add a new protocol. If people want the tracing, I'd happy to create an issue for it. 

If you want to know about the feature of the distributed tracing, you can refer this PR. You can find the documentation, video, and samples are there. It is now on review, so the contents could change, however, it might be a good place what it looks like. https://github.com/Azure/azure-functions-durable-extension/pull/1298

**Comment by MedAnd** [comment]
Hi @TsuyoshiUshio, thank you & will take a look at the PR documentation, video, and samples... 

I do note the following are not supported scenarios and wondering when they will be supported: 

Durable Entitry Orchestration (not supported)
HTTP endpoints (not supported)
Human Interaction (not supported)

Above is directly related to some feedback I shared here: https://github.com/Azure/azure-functions-durable-extension/issues/1305

**Comment by TsuyoshiUshio** [comment]
Hi @MedAnd 
Thank you for the questions and comment!  The not supported one is just not implemented. :) Not technical blocker. However, I'd like to implement one by one. These stories requires additional change for the other part rather than core implementation. 

## Raise event with Distributed Tracing

One exception is Human Interaction also, RaiseEvent one. This one requires discussion. Maybe no one doing this trace before. (or just I don't know.)  This scenario required multiple parent. The correlation system (Activity) assumes one parent. However, RaiseEvent (or Human interaction) could wait several event happens. That means, they could have multiple parent. In this case we need to discuss how to handle it. **Do you have any idea, it is very welcome!** 

Currently I have some idea for this scenario. 

## Idea of Raise event correlation 

We have 
1. Orchestration, wait for wait for event coming.   We have two parent(start orchestration and event raise. It has OperationId 1, 2 respectively.) 
2.  Assume the parent is the orchestration one. However, when we receive the event on the orchestrator, it will create a custom telemetry as a child of orchestration(1), that has a property to point the Id coming from raise event source. At the same time, create another custom telemetry that as a child of raise event(2) that has a property to point the Id coming from orchestration.
3.  Discuss with Azure Portal team to enable the link. 

However, I haven't discussed with anyone. Just an idea. So I'm going to discuss how to achieve this.  What do you think, @cgillum @anthonychu ?

**Comment by TsuyoshiUshio** [comment]
Sorry for the late response, however, I released pre-release version. For more details, you can find the announcement, video, documentation, samples and getting started! 
https://medium.com/@tsuyoshiushio/durable-functions-distributed-tracing-71426fe2246f

**Comment by cgillum** [comment]
Now that we have the alpha release done, let's use this issue to track a public preview release that is enabled by default.

**Comment by TsuyoshiUshio** [comment]
Hi @olitomlinson 
Sorry for late reply. I noticed that in case of Event Raised, the tracing is broken. I wrote some fix for it.  
https://github.com/Azure/durabletask/pull/430
Also we need to discuss how to correlate the raise event cases. It has two parents.  My current idea is, the primary parent(orchestration client call) will be the parent, and raise event id will be stored as a property.

**Comment by santo2** [comment]
is this expected to be worked on soon? We were waiting for it but we see that it hopped to the 2.4.0 version :-)

**Comment by davidmrdavid** [comment]
@santo2 , I'm actively working on it now! I started working on it just yesterday. We're first looking to merge the correlation feature branch into master in DurableTask, which you can track here: https://github.com/Azure/durabletask/pull/422

After that, we'll begin doing some validation of correctness and prioritizing the missing bits. So, in short, it's happening _now_!

⚡ ⚡

**Comment by davidmrdavid** [comment]
https://github.com/Azure/durabletask/pull/422 has been merged into DTFx! That doesn't mean we're done, but it's an important step!

**Comment by davidmrdavid** [comment]
As a follow-up, we just merged the correlation bits for `durable-extension` into the `dev` branch. As seen here: https://github.com/Azure/azure-functions-durable-extension/pull/1571

We'll provide more details about this in the upcoming release notes :) .

**Comment by santo2** [comment]
Hi @davidmrdavid ,
Thanks for the update! What is the easiest way to also get notified on the release of that merge?

**Comment by davidmrdavid** [comment]
Hi @santo2

The _easiest_ would be click the "Watch" button in this repo, and selecting a "custom" notification scheme where you make sure you select to get notifications for releases. That should give you an update every time we update this page: https://github.com/Azure/azure-functions-durable-extension/releases

**Comment by NickSevens** [comment]
@davidmrdavid with the 2.4.0 package, is the tracing still supposed to show under `DTOrchestration` and `DTActivity` labels?

**Comment by NickSevens** [comment]
It's working very well with the 2.4.0 package!

I do have a question: I'm performing some calls to external APIs in my functions, but they do not show up as a dependency in the "orchestration structure". Is there a way to also include calls (from HttpClient for example) in the entire dependency tree? Would be nice to see the complete picture in 1 go.

**Comment by ConnorMcMahon** [comment]
I don't believe we currently support that @NickSevens. I think that we may already support thatin our `IDurableOrchestrationContext.CallHttpAsync()` API, but adding some easier way to support that in activity functions may be a good idea.

I believe the best thing to do for now is to wrap those external HTTP calls in a custom service using [dependency injection](https://docs.microsoft.com/en-us/azure/azure-functions/functions-dotnet-dependency-injection#register-services). You can then have this custom service construct a telemetry client and [manually trace](https://github.com/Azure/azure-functions-durable-extension/blob/dev/samples/correlation-csharp/FunctionAppCorrelation/Monitoring.cs#L58) the HTTP request yourself.

If you have any ideas for how to more easily support this, it would probably be best to track that as a separate issue. The more details you put in the proposal, the more we have to work with as a development team to decide if and how we implement that functionality.

**Comment by ConnorMcMahon** [comment]
Closing the issue since this is now out officially in v2.4.0. I am sure there is still room for improvements, but we can track those issues as they come in.

**Comment by NickSevens** [comment]
Problem there is I'm using the Graph SDK for example, or SharePoint CSOM libraries, which do use HttpClient internally. Not sure how to implement custom logging in there... 😊

**Comment by ConnorMcMahon** [comment]
Interesting...

Unfortunately, it really depends on how those libraries retrieve their HTTP clients, so we may not have a way to hijack their http clients to add the telemetry ourselves.

I would still file an issue, because App Insights SDK may provide an extension point where we can intercept their outbound dependency tracking logic to add correlation automatically without the need to use a custom `HttpClient`.

**Comment by tommck** [comment]
Is this still in alpha after all this time?

**Comment by Shivam60** [comment]
Is this still in alpha?

**Comment by cgillum** [comment]
Yes, this is still in alpha. We made the decision to revisit the implementation to make it more compatible with the [alternate storage providers](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-storage-providers) (the current implementation is tightly coupled to Azure Storage) and to make it compatible with [OpenTelemetry](https://opentelemetry.io/) (the current prototype is tightly coupled to Application Insights).

The work to update the implementation is currently in-progress (see https://github.com/Azure/durabletask/pull/648), but the team hasn't yet committed to an ETA.

FYI @bachuv and @AnatoliB

**Comment by olitomlinson** [comment]
> the current implementation is tightly coupled to Azure Storage) and to make it compatible with OpenTelemetry (the current prototype is tightly coupled to Application Insights)

Great move!

**Comment by santo2** [comment]
we have been waiting for a looooong time ;-) I revisit this topic each month in the hopes that it's done!

**Comment by kepikoi** [comment]
We have also been waiting months if not years for this and the lack of trace correlation is still a serious problem. Makes debugging of custom (winston) logs in durable activities rocket science

**Comment by oising** [comment]
> Yes, this is still in alpha. We made the decision to revisit the implementation to make it more compatible with the [alternate storage providers](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-storage-providers) (the current implementation is tightly coupled to Azure Storage) and to make it compatible with [OpenTelemetry](https://opentelemetry.io/) (the current prototype is tightly coupled to Application Insights).
> 
> The work to update the implementation is currently in-progress (see [Azure/durabletask#648](https://github.com/Azure/durabletask/pull/648)), but the team hasn't yet committed to an ETA.
> 
> FYI @bachuv and @AnatoliB

Go Chris and team :) We know you're doing the right thing here, and it's painful to undo the appinsights stuff, but OT is The Way. Thank you!

**Comment by marnilss** [comment]
How can this issue be closed, when apparently feature is still is in alpha and you are working on it (see comments above)?
It is so difficult for us, end-users, of libraries to know what is supposed to work and not. I've been investigating and troubleshooting my own stuff because end-to-end distributed tracing does not work, and this issue was closed with text: "Closing the issue since this is now out officially in v2.4.0. I am sure there is still room for improvements, but we can track those issues as they come in."
And after reading the comments (posted after it was closed) I find out that it never reached GA, but is still in alpha state.

I think the documenation of Azure Functions, Durable Functions etc. is very hard to get a grip on. I've worked a lot more with ASP.NET Core and their documentation is in one place (docs....) and up to date. For functions you need to look at docs..., blog-articles and read all the comments in github-issues.

**Comment by dcbrown16** [comment]
Hi team, we have a customer who is calling an Angular single-page application. In their app, the client calls an HTTP Trigger Function, which calls an Orchestrator, which calls an Activity. They want to get the same operation_id across all of these Functions in data collected to Application Insights.
The latest progress thread linked (https://github.com/Azure/durabletask/pull/648) points to a .NET library (https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/README.md). Will Angular/JavaScript also be supported by this feature at some point?

**Comment by cgillum** [comment]
Yes, this should work fine as long as the Angular client includes the necessary distributed tracing headers in the HTTP request that triggers the HTTP function.
